File: LeetCode/cpp/1189_Maximum_Number_of_Balloons.cpp
class Solution {
public:
	int maxNumberOfBalloons(string text) {
		int answer = INT_MAX;

		unordered_map<char, int> um;

		for (int i = 0; i < text.length(); i++) {
			um[text[i]]++;
		}

		char alphabet[] = { 'b', 'a', 'l', 'o', 'n' };
		for (char c : alphabet) {
			if (um.count(c) == 0) {
				return 0;
			}
		}

		for (auto m : um) {
			if (m.first == 'b' || m.first == 'a' || m.first == 'n') {
				answer = min(answer, m.second);
				continue;
			}

			if (m.first == 'l' || m.first == 'o') {
				if (m.second < 2) {
					return 0;
				}

				answer = min(answer, m.second / 2);
			}
		}

		return answer;
	}
};

File: LeetCode/java/1189_Maximum_Number_of_Balloons.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int maxNumberOfBalloons(String text) {
        int answer = Integer.MAX_VALUE;

        Map<Character, Integer> map = new HashMap<>();

        for (int i = 0; i < text.length(); i++) {
            map.put(text.charAt(i), map.getOrDefault(text.charAt(i), 0) + 1);
        }

        char[] alphabet = {'b', 'a', 'l', 'o', 'n'};
        for (char c : alphabet) {
            if (!map.containsKey(c)) {
                return 0;
            }
        }

        for (Map.Entry<Character, Integer> m : map.entrySet()) {
            if (m.getKey() == 'b' || m.getKey() == 'a' || m.getKey() == 'n') {
                answer = Math.min(answer, m.getValue());
                continue;
            }

            if (m.getKey() == 'l' || m.getKey() == 'o') {
                if (m.getValue() < 2) {
                    return 0;
                }

                answer = Math.min(answer, m.getValue() / 2);
            }
        }

        return answer;
    }
}